cmd/ctl/pkg/status/certificate/types.go

```diff
@@ -0,0 +1,410 @@
+/*
+Copyright 2020 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificate
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strings"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubectl/pkg/describe"
+
+	"github.com/jetstack/cert-manager/cmd/ctl/pkg/status/util"
+	cmapiv1alpha2 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+type CertificateStatus struct {
+	// Name of the Certificate resource
+	Name string
+	// Namespace of the Certificate resource
+	Namespace string
+	// Creation Time of Certificate resource
+	CreationTime metav1.Time
+	// Conditions of Certificate resource
+	Conditions []cmapiv1alpha2.CertificateCondition
+	// DNS Names of Certificate resource
+	DNSNames []string
+	// Events of Certificate resource
+	Events *v1.EventList
+	// Not Before of Certificate resource
+	NotBefore *metav1.Time
+	// Not After of Certificate resource
+	NotAfter *metav1.Time
+	// Renewal Time of Certificate resource
+	RenewalTime *metav1.Time
+
+	IssuerStatus *IssuerStatus
+
+	SecretStatus *SecretStatus
+
+	CRStatus *CRStatus
+}
+
+type CertificateStatusBuilder struct {
+	// Name of the Certificate resource
+	Name string
+	// Namespace of the Certificate resource
+	Namespace string
+	// Creation Time of Certificate resource
+	CreationTime metav1.Time
+	// Conditions of Certificate resource
+	Conditions []cmapiv1alpha2.CertificateCondition
+	// DNS Names of Certificate resource
+	DNSNames []string
+	// Events of Certificate resource
+	Events *v1.EventList
+	// Not Before of Certificate resource
+	NotBefore *metav1.Time
+	// Not After of Certificate resource
+	NotAfter *metav1.Time
+	// Renewal Time of Certificate resource
+	RenewalTime *metav1.Time
```

Comment:
This is a really good change, but I find this bit strange.

We are storing almost a duplicate of `CertificateStatus` in the builder, when all
the other resources create a `Status` immediately.

I can think of a couple of ways of doing this that avoids all of this duplication:

1) immediately start filling in the `CertificateStatus` when calling the builder, rather than storing a copy and then copying in `.build()`.

2) Just have a `.Certificate` field on the builder, and then copy the info to the status in `.build()` rather than storing
the intermediate copy.

What do you think?